### PR TITLE
Add new arrangement 4

### PR DESCRIPTION
**Pull request by lcarva**
Signed-off-by: Luiz Carvalho <lucarval@redhat.com>

### CODE REVIEW

**Comment by lcarva** [comment]
Not sure if this is the best way to handle an in-flight arrangement version.

**Comment by mlangsdorf** [comment]
looks like it's missing post_group_manifest and exit_pulp_publish.

I think this code needs to wait until exit_pulp_publish has passed unit testing.

**Comment by lcarva** [comment]
@mlangsdorf right, it's missing all the plugins that don't really exist yet. That's why the default arrangement version wasn't incremented.

Although this is not needed, it would facilitate testing of these plugins as they're being added. Not sure how we want to handle that. Maybe something like `arrangement_version='dev'` ?

File: inputs/worker_inner:4.json [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        "kojihub": "{{KOJI_HUB}}",
+        "url": "{{OPENSHIFT_URI}}",
+        "verify_ssl": false,
+        "blocksize": 10485760,
+        "koji_keytab": false,
+        "koji_principal": false
+      }
+    },
+    {
+      "name": "pulp_pull",
```
Comment (vrutkovs) [inline inputs/worker_inner:4.json]: `pulp_pull` should run before `koji_upload` to make sure a correct schema 2 ID is stored

Comment (twaugh) [inline inputs/worker_inner:4.json]: It's `koji_import` which needs the correct ID.

`pulp_pull` should be moved to the orchestrator build.

**Comment by lcarva** [comment]
Updated to reflect what arrangement 4 will look like once all plugins are in place.

File: inputs/orchestrator_inner:4.json [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    },
+    {
+      "name": "koji_parent",
+      "required": false,
+      "args": {
+        "koji_hub": "{{KOJI_HUB}}"
+      }
+    },
+    {
+      "name": "reactor_config",
```
Comment (twaugh) [inline inputs/orchestrator_inner:4.json]: I'd rather this plugin ran as the first in the pre-build phase.

File: inputs/worker_inner:4.json [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        "command": "{{SOURCES_COMMAND}}"
+      }
+    },
+    {
+      "args": {
+        "koji_root": "{{KOJI_ROOT}}",
+        "koji_hub": "{{KOJI_HUB}}"
+      },
+      "required": false,
+      "name": "fetch_maven_artifacts"
```
Comment (twaugh) [inline inputs/worker_inner:4.json]: Move "name" to the top for easier reading?

File: inputs/worker_inner:4.json [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        }
+      }
+    },
+    {
+      "name": "pulp_push",
+      "args": {
+        "pulp_registry_name": "{{PULP_REGISTRY_NAME}}",
+        "load_exported_image": true,
+        "dockpulp_loglevel": "INFO",
+        "sync": false
```
Comment (twaugh) [inline inputs/worker_inner:4.json]: This should be "publish", not "sync" (my fault, I wrote it up wrong originally).
http://osbs.readthedocs.io/en/latest/multiarch.html#pulp-push

File: inputs/orchestrator_inner:4.json [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    },
+    {
+      "name": "pulp_tag",
+      "args": {
+      }
+    },
+    {
+      "name": "pulp_sync",
+      "args": {
+        "pulp_registry_name": "{{PULP_REGISTRY_NAME}}",
```
Comment (twaugh) [inline inputs/orchestrator_inner:4.json]: I'm now thinking we want a `"publish": false,` here as well as in pulp_push.
https://github.com/projectatomic/osbs-docs/pull/36

File: inputs/worker_inner:4.json [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+      "args": {
+        "labels": "{{IMPLICIT_LABELS}}"
+      }
+    },
+    {
+      "name": "change_from_in_dockerfile"
+    },
+    {
+      "name": "add_help",
+      "required": false
```
Comment (twaugh) [inline inputs/worker_inner:4.json]: Can drop `"required": false` now I think?

File: inputs/worker_inner:4.json [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+      "args": {
+        "command": "{{SOURCES_COMMAND}}"
+      }
+    },
+    {
+      "args": {
+        "koji_root": "{{KOJI_ROOT}}",
+        "koji_hub": "{{KOJI_HUB}}"
+      },
+      "required": false,
```
Comment (twaugh) [inline inputs/worker_inner:4.json]: Can drop `"required": false` now I think?

File: tests/build_/test_arrangements.py [hunk 8]
```diff
@@ -110,6 +111,10 @@ def assert_plugin_not_present(self, build_json, phase, name):
         with pytest.raises(NoSuchPluginException):
             get_plugin(plugins, phase, name)
 
+    def get_pulp_sync_registry(self, conf):
+        """Return the docker registry used by pulp content sync."""
+        return conf.get_registry_uris()[0].split('/')[0]
```
Comment (twaugh) [inline tests/build_/test_arrangements.py]: Should be the first *V2* registry.

File: tests/build_/test_arrangements.py [hunk 9]
```diff
@@ -667,3 +672,254 @@ def test_fetch_worker_metadata(self, osbs, scratch):
...
+    - tag_from_config enabled
+    - pulp_tag enabled
+    - pulp_sync enabled
+    - pulp_publish enabled
+    - pulp_pull enabled
+
+    Worker build differences from arrangement version 3:
+    - tag_from_config takes "tag_suffixes" argument
+    - tag_by_labels disabled
+    - pulp_push takes an additional "sync":false argument
```
Comment (twaugh) [inline tests/build_/test_arrangements.py]: This will need updating to "publish".

File: tests/build_/test_arrangements.py [hunk 10]
```diff
@@ -667,3 +672,254 @@ def test_fetch_worker_metadata(self, osbs, scratch):
...
+        unique_tag_suffix = ''
+        if has_plat_tag:
+            unique_tag_suffix = '-' + additional_params.get('platform')
+        assert unique_tags[0].endswith(unique_tag_suffix)
+
+        primary_tags = args['tag_suffixes']['primary']
+        if has_primary_tag:
+            assert set(primary_tags) == set(['latest', '{version}', '{version}-{release}'])
+
+    def test_tag_by_labels(self, osbs):  # noqa:F811
```
Comment (twaugh) [inline tests/build_/test_arrangements.py]: I think we don't need this test, as the "running order" test covers it.

**Comment by lcarva** [comment]
@twaugh, I think I addressed all of you comments.

@mlangsdorf, changes now include `pulp_publish` and `group_manifests`.

**Comment by lcarva** [comment]
Updated release notes.

**Pull request merged**